vue-components/src/components/Popover.vue

```diff
@@ -112,14 +129,173 @@ export default Vue.extend( {
+				[ PopoverPositions.START ]: PopoverPositions.TOPSTART,
+				[ PopoverPositions.STARTTOP ]: PopoverPositions.TOPSTART,
+				[ PopoverPositions.STARTBOTTOM ]: PopoverPositions.BOTTOMSTART,
+				[ PopoverPositions.END ]: PopoverPositions.TOPEND,
+				[ PopoverPositions.ENDTOP ]: PopoverPositions.TOPEND,
+				[ PopoverPositions.ENDBOTTOM ]: PopoverPositions.BOTTOMEND,
+			};
+			this.adjustedPosition = swappedPositions[ this.position ] || null;
+
+			await this.$nextTick();
+		},
+		/**
+		 * Try to avoid having the popover overflow the viewport,
+		 * by adjusting the horizontal subposition (start/center/end).
+		 *
+		 * For each horizontal subposition, this tries the other two options,
+		 * and picks the one out of the three with the least overflow.
+		 */
+		async avoidOverflowByAdjustingHorizontalSubposition(): Promise<void> {
+			const initialOverflow = this.getOverflow();
+			if ( !initialOverflow ) {
+				return;
+			}
+
+			let lowestOverflow = initialOverflow;
+			let bestAdjustedPosition = this.adjustedPosition;
+			const otherPositions: Partial<Record<string, string[]>> = {
+				[ PopoverPositions.TOPSTART ]: [ PopoverPositions.TOP, PopoverPositions.TOPEND ],
+				[ PopoverPositions.TOP ]: [ PopoverPositions.TOPSTART, PopoverPositions.TOPEND ],
+				[ PopoverPositions.TOPEND ]: [ PopoverPositions.TOP, PopoverPositions.TOPSTART ],
+				[ PopoverPositions.BOTTOMSTART ]: [ PopoverPositions.BOTTOM, PopoverPositions.BOTTOMEND ],
+				[ PopoverPositions.BOTTOM ]: [ PopoverPositions.BOTTOMSTART, PopoverPositions.BOTTOMEND ],
+				[ PopoverPositions.BOTTOMEND ]: [ PopoverPositions.BOTTOM, PopoverPositions.BOTTOMSTART ],
+			};
+			for ( const otherPosition of otherPositions[ this.adjustedPosition || this.position ] || [] ) {
+				this.adjustedPosition = otherPosition;
+				await this.$nextTick();
+				const overflowAfterAdjustment = this.getOverflow();
+				if ( overflowAfterAdjustment < lowestOverflow ) {
+					lowestOverflow = overflowAfterAdjustment;
+					bestAdjustedPosition = otherPosition;
+				}
+			}
+
+			if ( bestAdjustedPosition !== this.adjustedPosition ) {
+				this.adjustedPosition = bestAdjustedPosition;
+				await this.$nextTick();
+			}
+		},
+		/**
+		 * Try to avoid having the popover overflow the viewport,
+		 * by making its edges flush with those of the target button.
+		 *
+		 * By default, for the non-center horizontal subpositions,
+		 * both the pointer and the edge of the popover are close to the center of the button.
+		 * We can gain some space by moving the edge of the popover to the edge of the button
+		 * (assuming that the button has some substantial width:
+		 * for small buttons, especially icon buttons, this makes no difference),
+		 * at the cost of also moving the pointer along with it.
+		 * The movement of the pointer away from the button’s center is unwanted,
+		 * but unavoidable given the current DOM structure
+		 * (where it’s positioned relative to the popover, not the button).
```

Comment:
Demonstration of the difference from this function, including the unwanted pointer movement:
![center](https://user-images.githubusercontent.com/2346599/122417727-fdc5cd00-cf89-11eb-83fd-bc7831d08915.png)
![flush](https://user-images.githubusercontent.com/2346599/122417764-05857180-cf8a-11eb-9619-d03b7826b72b.png)
The unwanted pointer movement is the reason that this adjustment is relatively late in the list of adjustments – we try jiggling the position first, and if that works, then the pointer can stay centered on the button.